docs/conceptual/runtime/weight.md

```diff
@@ -0,0 +1,159 @@
+  prevent restrict certain dispatches through any means that makes sense to that particular chain.
+  An example of such dispatches in substrate is the call to a smart contract (todo: link when this
+  is implemented).
+
+The system module is responsible for accumulating the weight of each block as it gets executed and
+making sure that it does not exceed the limit. The transaction-payment module is responsible for
+interpreting these weights and deducting fees based upon them.
+
+
+## Block Weight and Length Limit
+
+Aside from affecting the fees, the main purpose of the weight system is to prevent a block from
+being too filled with transactions. The system module, while processing transactions within a block,
+accumulates both the total length of the block (sum of encoded transactions in number of bytes) and
+the total weight of the blocks. At any points, if these numbers surpass the limits, no further
+transactions are accepted to the block. These limits are defined in
+[`MaximumBlockLength`](/rustdocs/master/srml_system/trait.Trait.html#associatedtype.MaximumBlockLength)
+and
+[`MaximumBlockWeight`](/rustdocs/master/srml_system/trait.Trait.html#associatedtype.MaximumBlockLength)
+respectively.
+
+One important note about these limits is that a ratio of them are reserved for the `Operational`
+dispatch class. This rule applies to both of the limits and the ratio can be found in
+[`AvailableBlockRatio`](/rustdocs/master/srml_system/trait.Trait.html#associatedtype.AvailableBlockRatio).
+For example, if the block length limit is 1 mega bytes and the ratio is set the 80%, all
+transactions can fill the first 800 kilo bytes of the block while the last 200 can only be filled by
+the operation class.
+
+## Custom Weight Implementation
+
+Implementation a custom weight calculation function can vary from _very easy_ to _super
+complicated_. The `SimpleDispatchInfo` struct provided by substrate is fairly one of the easiest
+appraoches. Anything more sophisticated would require a bit more work.
+
+Any weight calculation function must provide two trait implementations:
+  - [`WeightData<T>`]: To determine the weight of the dispatch.
+  - [`ClassifyDispatch<T>`]: To determine the class of the dispatch. See the enum definition for
+    more information on dispatch classes. This struct can then be used with `#[weight]` annotation.
+
+Substrate then bundles then output information of the two traits into [`DispatchInfo`] struct and
+provides it by implementing the [`GetDispatchInfo`] for all `Call` variants, and opaque extrinsic
+types. This is used internally by the system and executive module; you won't use it most likely.
+
+Both `ClassifyDispatch` and `WeightData` are generic over `T` which gets resolved into the tuple of
+all dispatch arguments except for the origin. To demonstrate, we will craft a struct that calculates
+the weight as `m * len(args)` where `m` is a given multiplier and `args` is the concatenated tuple
```

Comment:
You get pretty heavy into implementation details here. I think this starts to become a developer doc at this point...